budgetportal/tests/test_provincial_infrastructure_projects.py

```diff
@@ -86,3 +87,88 @@ def test_upload_xlsx_for_prov_infra_projects(self):
         self.assertEqual(count, 11)
 
 
+class IRMReportSheetTestCase(TestCase):
+    def setUp(self):
+        dataset = Dataset()
+        dataset.headers = NORMAL_HEADERS + ["Project Contractor"] * 50
+        row = (
+            [1] * len(NORMAL_HEADERS)
+            + [
+                "Program Implementing Agent: DOPW",
+                "Program Implementing Agent: TEST",
+                "Main Contractor: AAAA",
+                "Principal Agent: BBBB",
+            ]
+            + [None] * 46
+        )
+        dataset.append(row=row)
+        self.report_sheet = IRMReportSheet(dataset)
+
+        dataset_with_other_keys = Dataset()
+        dataset_with_other_keys.headers = NORMAL_HEADERS + ["Project Contractor"] * 50
+        row_with_other_keys = (
+            [1] * len(NORMAL_HEADERS)
+            + [
+                "Service Provider: DOPW",
+                "Program Implementing Agent: TEST",
+                "Service Provider: AAAA",
+                "Principal Agent: BBBB",
+            ]
+            + [None] * 46
+        )
+        dataset_with_other_keys.append(row_with_other_keys)
+        self.sheet_with_other_keys = IRMReportSheet(dataset_with_other_keys)
+
+    def test_clean_project_contractors(self):
+        """Check that first dataset has 50 Project Contractor columns but output has 0"""
+
+        report_sheet = self.report_sheet
+        num_project_contractor_columns = len(report_sheet.contractor_columns)
+        self.assertEqual(num_project_contractor_columns, 50)
+
+        report_sheet.process()
+        new_report_sheet = IRMReportSheet(report_sheet.output_data_set)
+
+        num_project_contractor_columns = len(new_report_sheet.contractor_columns)
+        self.assertEqual(num_project_contractor_columns, 0)
+
+    def test_assigned_correctly(self):
+        """Check that project contractors successfully mapped"""
+
+        self.report_sheet.process()
+        program_implementing_agent = self.report_sheet.output_data_set[
+            "Program Implementing Agent"
+        ]
+        main_contractor = self.report_sheet.output_data_set["Main Contractor"]
+        principal_agent = self.report_sheet.output_data_set["Principal Agent"]
+        other_parties = self.report_sheet.output_data_set["Other parties"]
+
+        self.assertEqual(program_implementing_agent, ["DOPW\nTEST"])
+        self.assertEqual(main_contractor, ["AAAA"])
+        self.assertEqual(principal_agent, ["BBBB"])
+        self.assertEqual(other_parties, [None])
+
+    def test_other_parties(self):
+        """Checks when different keys given in project contractor"""
+
+        # it basically does the same what process()->process_row() does
+        for index, row in enumerate(self.sheet_with_other_keys.data_set):
```

Comment:
why is this happening? it looks hacky - isn't there a more normal way of making the processing happen?

Reply:
That will also improve data locality - where the test data is defined and checked will be close together, and not interleaved with other tests

Reply:
changed the loop with process() method
https://github.com/vulekamali/datamanager/pull/189/commits/2a6005d6eec5ea47e4d404aeeb05b287fbd797ce